petastorm/pytorch.py

```diff
@@ -348,7 +365,38 @@ def _instantiate_noop_buffer():
 
         return shuffling_buffer, other_shuffling_buffer, instantiate_buffer_fn
 
+    def _worker(self):
+        for b in self._iter_impl_worker():
+            if self._finished_event.is_set():
+                break
+            self._async_batch_queue.put(b)
+        else:
+            self._async_batch_queue.put(None)
+
     def _iter_impl(self):
+        if self._async_batch_queue:
+            # Get shuffled batches from async shuffle thread
+            if not self._started:
+                # If reuse dataloader in a new pass, async shuffle thread has to be relaunched.
+                # It is finished in previous pass when reaching end of data.
+                self._shuffle_thread = Thread(target=self._worker)
+                self._shuffle_thread.daemon = True
+                self._started = True
+                self._shuffle_thread.start()
+            while True:
+                b = self._async_batch_queue.get()
+                if b is None:
+                    # None is last data in async batch queue, which means that async shuffle thread
+                    # has reached the end of data.
+                    self._started = False
```

Comment:
Should we join here (and maybe self._shuffle_thread=None) for the sake of leaving a clean state on exit? Should we actually use `_close_async_thread` here?